### PR TITLE
Fix bug with texture coordinate q value being 0

### PR DESCRIPTION
**Pull request by OrN**
This exists only to get fifoci diffs and for JMC47 to play with for now.

[Review on Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3707)

### CODE REVIEW

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by JMC47** [comment]
The Lesson08 fifolog more or less shows what's going on with this.

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild

**Comment by JMC47** [comment]
This is the difference in the NSMBWii coins is a lie.

damn

**Comment by JMC47** [comment]
This is the most heavily tested change in the history of Dolphin. 50+ dols, multiple fifologs, screenshots vs hardware, and tons of crazy stuff.

The behavior doesn't affect any games, but fixes Lesson08 and a whole bevy of other texture matrix issues that games don't seem to care about.

**Comment by AdmiralCurtiss** [comment]
So what is actually going on with this PR?

**Comment by JMC47** [comment]
Hardest part about this: finding a fucking game that used it.  I fucking told you.  Some developers would be stupid enough to use it.

Fixes HB Disc Golf and HB Arcade Cards

**Comment by JMC47** [comment]
HB Arcade Cards
Master
wlgell-2
PR
wlgell-1

**Comment by OrN** [comment]
Lesson08 is the only true case.

**Comment by JMC47** [comment]
For Metroid Prime 3, the lighting is much, much more accurate on this PR.

metroidprime3consolelesson08

**Comment by OrN** [comment]
That may not be correct

**Comment by JMC47** [comment]
It figures that upon finally figuring out every last fucking case all the sudden a Factor 5 game is affected.

**Comment by BhaaLseN** [comment]
Do you have a useful comment to put next to the `z == 0` changes? I'm concerned that if we don't comment it properly now that we might run into surprises like the special condition in those other PRs.

**Comment by magcius** [comment]
So, for context, STQ are texture coordinates. Q is the perspective divide factor, similar to W in XYZW. Since the Wii is right-handed, Q>=0 and W>=0 are undefined cases (the point is behind the camera, or is on the camera plane itself). However, "undefined" in this case means "whatever the hardware does in this undefined edge case". In this case, the hardware seems to divide the resulting texture coordinates by 2. This should not be relied on by any developers, but....

File: Source/Core/VideoBackends/Software/SWVertexLoader.cpp [hunk 1]
```diff
@@ -234,6 +234,9 @@ void SWVertexLoader::ParseVertex(const PortableVertexDeclaration& vdec, int inde
 	{
 		ReadVertexAttribute<float>(m_Vertex.texCoords[i], src, vdec.texcoords[i], 0, 2, false);
 
+		// tex coord 'q' should be set to 1 initially, even if input form is XF_TEXINPUT_ABC1
+		m_Vertex.texCoords[i][2] = 1.0f;
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/SWVertexLoader.cpp]: Oh crap, was this a out-of-bounds memory access?

**Comment by BhaaLseN** [comment]
Sigh, switching to reviewable since keeping track is getting a little harder :P

---

Reviewed 1 of 2 files at r20, 2 of 3 files at r22.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/VideoCommon/VertexShaderGen.cpp, line 309 \[r22\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3707#-KE2IQ8fFNvaJiowL9lH:-KE2IQ8fFNvaJiowL9lI:757643669) ([raw file](https://github.com/dolphin-emu/dolphin/blob/5dde6235c0d03686fe753ae60873cb52dfb47348/Source/Core/VideoCommon/VertexShaderGen.cpp#L309)):*
Space after the `if`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3707#-:-KE2IbkaAupkf1qg0u1w:1148270684)*

**Comment by degasus** [comment]
Reviewed 1 of 3 files at r22, 1 of 1 files at r23.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[Source/Core/VideoBackends/Software/TransformUnit.cpp, line 197 \[r23\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3707#-KE5U35zHudd0aNYsaWf:-KE5U36-F-oQOQ0HxCZc:-261467541) ([raw file](https://github.com/dolphin-emu/dolphin/blob/68ac65c494794c6ae60aed1510c7fbc4940144b7/Source/Core/VideoBackends/Software/TransformUnit.cpp#L197)):*
By the way, where is the else clause? I miss the division. This z==0 clause should be close to the division IMO

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3707)*

**Comment by JMC47** [comment]
Can this be rebased now that the feature freeze is over?

**Comment by JMC47** [comment]
Documenting this here because I remember us messing with Fishing Resort a bit - 

"<HdkR> Speaking of the Fishing Resort Wii map is inverted thing
<HdkR> A while ago I looked in to that, and it looks like it is doing a perspective divide on the texture coordinates with the perspective divide value being 0.0 "

**Comment by OrN** [comment]
I don't have decent enough internet atm to clone and rebase/modify atm

**Comment by JMC47** [comment]
Can anyone review this.  I've retested it and it fixes 4 games while breaking nothing I know about.

**Comment by phire** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r26.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3707#-:-KQXutsd2KRySzTF1QuX:bnfp4nl)*

**Comment by JMC47** [comment]
I reviewed the code comments to make sure they were accurate and such.  I also did the hw testing and have like 50 dols that prove this is accurate to hardware.

**Pull request merged**